nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListenFTP.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @OnScheduled
+    public void startFtpServer(ProcessContext context) {
+        if (ftpServer == null) {
+            String username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+            String password = context.getProperty(PASSWORD).evaluateAttributeExpressions().getValue();
+            String bindAddress = context.getProperty(BIND_ADDRESS).evaluateAttributeExpressions().getValue();
+            int port = context.getProperty(PORT).evaluateAttributeExpressions().asInteger();
+
+            try {
+                sessionFactorySetSignal = new CountDownLatch(1);
+                ftpServer = new NifiFtpServer.Builder()
+                        .sessionFactory(sessionFactory)
+                        .sessionFactorySetSignal(sessionFactorySetSignal)
+                        .bindAddress(bindAddress)
+                        .port(port)
+                        .username(username)
+                        .password(password)
+                        .build();
+                ftpServer.start();
+            } catch (ProcessException processException) {
+                getLogger().error(processException.getMessage(), processException);
+                stopFtpServer();
+                throw processException;
+            }
+        } else {
+            getLogger().warn("Ftp server already started.");
+        }
+    }
+
+    @OnStopped
+    public void stopFtpServer() {
+        if (ftpServer != null && !ftpServer.isStopped()) {
+            ftpServer.stop();
+        }
+        ftpServer = null;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+        if (this.sessionFactory.compareAndSet(null, sessionFactory)) {
+            sessionFactorySetSignal.countDown();
+        }
+        context.yield();
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext context) {
+        List<ValidationResult> results = new ArrayList<>(2);
+        String username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        String password = context.getProperty(PASSWORD).evaluateAttributeExpressions().getValue();
+
+        if ((username == null) && (password != null)) {
+            results.add(usernameOrPasswordIsNull(USERNAME, PASSWORD));
+        } else if ((username != null) && (password == null)) {
+            results.add(usernameOrPasswordIsNull(PASSWORD, USERNAME));
+        } else if ((username != null) && (password != null)) {
+            validateAgainstEmptyString(username, USERNAME, results);
+            validateAgainstEmptyString(password, PASSWORD, results);
+        }
+        return results;
+    }
+
+    private ValidationResult usernameOrPasswordIsNull(PropertyDescriptor nullProperty, PropertyDescriptor nonNullProperty) {
```

Comment:
Actually, this is more explicit this way when seen on the UI because it points out which property is left blank. In the ValidationResult, the problematic property is provided (the one that is null in this case), and the framework appends this error message to that. So when the username is left blank, it says "'Username' is invalid because 'Username' and 'Password' should either both be provided or none of them", and if the password is left blank, it says "'Password' is invalid because 'Password' and 'Username' should either both be provided or none of them". The variable order is a small trade-off for this piece of code to be reusable.